### DISCUSSION TITLE
How to load images from remote shared folder location in local network

### DISCUSSION THREAD

**Discussion by roshtha** [libvips/libvips#2768] [Q&A]
Hello,

I developed a C++ application in Windows platform using `libvips` DLLs for merging and creating a single pyramidal tiff file from a set of microscopy images. 
Currently, the images are accessed from a local folder in system. The code used `vips_image_new_from_file` for image loading.

```
int main(int argc, char **argv)
{
        int numrows = 54 ;
        int numcols = 16;
        int numimages = numrows * numcols;
        std::string tr ="";
        std::string tc ="";
        std::string fname = "";
        std::string outfilename = "out.tiff";
        int indexCount = 0;


        VipsImage *inImages[numimages];
        VipsImage *result;

        if( VIPS_INIT( argv[0] ) )
            vips_error_exit( NULL );

        vips_leak_set( TRUE );

       //Dynamically forms jpg source image names
        for (int r = 0; r < numrows; r++)
        {
            if( r < 9 )
            {
                tr = "0" + std::to_string(r+1);
            }
            else
            {
                tr = "" + std::to_string(r+1);
            }
            for (int c = 0; c < numcols; c++)
            {
                if( c < 9 )
                {
                    tc = "0" + std::to_string(c+1);
                }
                else
                {
                    tc = "" + std::to_string(c+1);
                }
                fname = std::string("ImageName_r00") + std::string(tr);
                fname.append("_c00");
                fname.append(tc);
                fname.append(".jpg");

                if( !(inImages[indexCount] = vips_image_new_from_file( fname.c_str(), "access", VIPS_ACCESS_SEQUENTIAL,
                                                                        NULL )) )
                {
                    vips_error_exit( NULL );
                }

                indexCount = indexCount + 1;
            }

        }
        auto start = high_resolution_clock::now();

        if( vips_arrayjoin(  inImages, &result, numimages, "across", numcols, NULL))
        {
            vips_error_exit( NULL );
        }
       for(int i = 0; i < numimages ; i++ )
            g_object_unref( inImages[i] );
        //jpeg_compression_75
        if( vips_tiffsave( result, outfilename.c_str(), "compression", VIPS_FOREIGN_TIFF_COMPRESSION_JPEG, "Q", 75, "tile-width", 
            256,"tile-height", 256,"pyramid", true, "tile", true, NULL ) )
            vips_error_exit( NULL );
       std::cout << "Finished creating pyramidal tiles for WSI" <<std::endl;
        g_object_unref( result );

        auto stop = high_resolution_clock::now();
        auto duration = duration_cast<minutes>(stop - start);

        cout << "Time taken for the process:" << duration.count() << " minutes" << endl;
        return 0;
}
```
Now I want to access images from a remote network shared folder in LAN which also requires credentials to access. 
How can I load images from a shared network folder using `libvips`  with options to specify credentials?

Thanks.

**Comment by jcupitt** [comment]
Hi @roshtha,

You can either map a network drive, or you could subclass `VipsSource`, adding open and read methods that handle these extra features.

**Comment by jcupitt** [comment]
Here's a sample program with a custom source.

```C
/* compile with:
 *      gcc -g -Wall customsource.c `pkg-config vips --cflags --libs`
 */

#include <vips/vips.h>

typedef struct _FileState {
  FILE *handle;
  guint counter;
} FileState;

static gint64 my_read(VipsSourceCustom *source, void *buffer, gint64 length,
                      void *user) {
  FileState *state = (FileState *)user;
  gint64 read;

  if (state->counter == 10) {
    return 0;
  } else {
    read = fread(buffer, sizeof(char), length, state->handle);
  }

  printf("read: %ld\n\r", read);

  state->counter = state->counter + 1;
  return read;
}

static gint64 my_seek(VipsSourceCustom *source, gint64 pos, int whence,
                      void *user) {
  return -1;
}

int main(int argc, char **argv) {
  VipsSourceCustom *source;
  VipsImage *image;
  FileState state;

  if (VIPS_INIT(argv[0]))
    vips_error_exit(NULL);

  source = vips_source_custom_new();
  if (!(state.handle = fopen(argv[1], "rb")))
    vips_error_exit("unable to open %s", argv[1]);

  state.counter = 0;

  g_signal_connect(source, "read", G_CALLBACK(my_read), &state);
  g_signal_connect(source, "seek", G_CALLBACK(my_seek), &state);

  image = vips_image_new_from_source(VIPS_SOURCE(source), "",
                  "access", VIPS_ACCESS_SEQUENTIAL,
                  "fail", TRUE,
                  NULL);
  if (!image)
    vips_error_exit(NULL);

  if (vips_image_write_to_file(image, "out.jpg", NULL))
    vips_error_exit(NULL);

  printf("successful write\n");

  g_object_unref(image);
  g_object_unref(source);
  fclose(state.handle);

  return 0;
}
```

That's using the signals on `VipsCustomSource`. You can also just subclass `VipsSource`, but that needs some knowledge of GObject.

**Reply by kmartinez** [reply]
Hi - if its windows and you can see the network drive in the file browser - you're best bet as John says is to map a drive (in Win10) then just use that path. You will gain a bit of caching too - which may speed things up.

**Reply by roshtha** [reply]
Hi @jcupitt and @kmartinez ,

Thanks for the hints.
I could use the following code to make it work from a remote image folder in Windows 10. 

     DWORD dwRetVal;

     NETRESOURCE nr;
     DWORD dwFlags;
     wchar_t pswd[] = L"user";
     wchar_t usrnm[] = L"pass";
     wchar_t remoteName[] = L"\\\\<ipadress>\\ScannerImages";
     memset(&nr, 0, sizeof (NETRESOURCE));
     nr.dwType = RESOURCETYPE_ANY;
     nr.lpLocalName = NULL; 
     nr.lpRemoteName = remoteName; 
     nr.lpProvider = NULL;
     dwFlags = CONNECT_UPDATE_PROFILE;
     dwRetVal = WNetAddConnection2(&nr, pswd, usrnm, dwFlags);
     if (dwRetVal == NO_ERROR)
            {
                wprintf(L"Connection added to %s\n", nr.lpRemoteName);
                sfile="\\\\<ipaddress>\\ScannerImages\\";
            }
            else
            {
                wprintf(L"WNetAddConnection2 to remote location failed with error: %u\n", dwRetVal);
            }

Then modified `fname` to include remote folder location `sfile`.

```
                fname = sfile;
                
                fname.append(std::string("ImageName_r00") + std::string(tr));
                fname.append("_c00");
                fname.append(tc);
                fname.append(".jpg");
```

Thanks.

**Comment by ohliumliu** [comment]
Hello,
   I am facing a similar situation -- trying to do arrayjoin usign pyvips to stitch a number of images. When the images are on local hard drive (Windows 10), it works like a charm. When I use a Cloudberry drive (AWS s3 bucket mapped as a Windows drive), it became very slow. For example, here is the output when running the script to stitch images from a mapped drive
```
stitch_images_pyvips_array.py temp-81: 120384 x 36480 pixels, 1 threads, 120384 x 1 tiles, 128 lines in buffer
stitch_images_pyvips_array.py temp-81: 14% complete
```
It took more than 30 min to go from 10% to 14% complete. Checking network traffic, download speed is only about a couple Mbps -- much lower than the available bandwidth of at least 50 Mbps.

When using local drive, the whole process took about 1 min.

I wonder if my vips setting is off when running the script. Your suggestion will be much appreciated!

**Reply by jcupitt** [reply]
Maybe try `access="random"` to `new_from_file`? It might help.

**Reply by ohliumliu** [reply]
Thank you for the suggestion!

Here is a little more background. I am looping through the coordinates and make a list of vips images. Some of them are images made with `new_from_file` while others are simply black images. It looks like the following:

```
    images = []
    for x in range(10):
        for y in range(100):
                if is_image(x,y):
                    images.append(pyvips.Image.new_from_file('path_to_image'), access = "random"))
                else:
                    images.append(pyvips.Image.black(width, height))
    full_image = pyvips.Image.arrayjoin(images, across = 10)
    full_image.write_to_file(f'''{os.path.expanduser("~")}/{outname}''', 
                             bitdepth = 8, xres = 100, yres = 100, compression = 'lzw')
```

I tried both "random" and "sequential". And the result is similar, download speed is much smaller than expected. It also showed 128 lines in buffer (sequential access) or 256 lines in buffer (random access). Random access did help a bit. It was slow at the beginning, and after about 10 min, it became much faster -- download speed is around 100 Mbps. I tried to increase VIPS_DISC_THRESHOLD from 100 to 100000, and it is about the same.

Still what is nice about this approach is that it spent almost no time saving the full image to local drive (the last step). I guess this is expected since vips optimizes the workflow.

The fastest approach I have found so far is to use `insert` to position the images made by `new_from_array` onto a black canvas. It works very well -- about 1/3 memory usage compared with a naiive use of numpy and about 2-3x faster. This is already very satisfactory for my use case, but it would be even better if I can make full use of vips library's features.
